core/net/mac/tsch/tsch.c

```diff
@@ -609,33 +609,37 @@ PT_THREAD(tsch_scan(struct pt *pt))
 
   static struct input_packet input_eb;
   static struct etimer scan_timer;
+  /* Time when to switch to the next channel */
+  static clock_time_t next_channel_time;
 
   ASN_INIT(current_asn, 0, 0);
 
   etimer_set(&scan_timer, CLOCK_SECOND / TSCH_ASSOCIATION_POLL_FREQUENCY);
 
+  next_channel_time = clock_time() + TSCH_ASSOCIATION_CHANNEL_SCAN_DURATION;
+
   while(!tsch_is_associated && !tsch_is_coordinator) {
     /* Hop to any channel offset */
     static int current_channel = 0;
-    /* Time when we started scanning on current_channel */
-    static clock_time_t current_channel_since = 0;
 
     /* We are not coordinator, try to associate */
     rtimer_clock_t t0;
     int is_packet_pending = 0;
-    clock_time_t now_seconds = clock_seconds();
+    clock_time_t now_time = clock_time();
 
     /* Switch to a (new) channel for scanning */
-    if(current_channel == 0 || now_seconds != current_channel_since) {
+    if(current_channel == 0 || CLOCK_LT(next_channel_time, now_time)) {
       /* Pick a channel at random in TSCH_JOIN_HOPPING_SEQUENCE */
       uint8_t scan_channel = TSCH_JOIN_HOPPING_SEQUENCE[
-          random_rand() % sizeof(TSCH_JOIN_HOPPING_SEQUENCE)];
+          (random_rand() >> 2) % sizeof(TSCH_JOIN_HOPPING_SEQUENCE)];
       if(current_channel != scan_channel) {
         NETSTACK_RADIO.set_value(RADIO_PARAM_CHANNEL, scan_channel);
         current_channel = scan_channel;
         PRINTF("TSCH: scanning on channel %u\n", scan_channel);
       }
-      current_channel_since = now_seconds;
+      /* select a random interval of time to spend scanning the channel */
+      next_channel_time = now_time + TSCH_ASSOCIATION_CHANNEL_SCAN_DURATION / 8
+          + (random_rand() >> 2) % (TSCH_ASSOCIATION_CHANNEL_SCAN_DURATION);
```

Comment:
I'm no fan of `random_rand() >> 2`, we already had a similar thing in `tsch-queue.c` and I was not happy about it, now making this more general doesn't feel good. Shouldn't we rather fix `random_rand` for platforms that have the problem?
As for getting rid of `CLOCK_LT`, would it work to add a variable `next_scan_duration`, and do `now_time - last_time > next_scan_duration`? I think this works regardless of the type because we now that `last_time` is in the past, that is, semantically lower than `now_time` (may be higher if there is wrap but means an earlier clock).